feat(ses): finite deep stacks, on by default

erights opened the pull request:
The deep stacks continue to use the `assert.note` mechanism to associate an error with previous stack-capturing errors. The unbounded leak that caused us to turn deep stacks off was really the unbounded memory of error-note-graphs.

Prior to this PR, the line disabling deep stacks by default was commented
```js
// ... We intend to change the default after verifying that having
// the feature enabled in production does not cause memory to leak.
```

This PR makes the error-note storage into a finite budget LRU cache, defaulting to a budget of `1000`. With the memory leak fixed, `track-turns.js` is changed to enable deep stacks by default. Measurements will still be needed to see what the cost is, and what is a reasonable budget.

erights commented:
@kriskowal if you could have a look at this, that'd be great. CI is giving me a jackpot of errors, all of which seem rooted in

```
Error running main: TypeError: Cannot read property 'length' of undefined
    at main (file:///home/runner/work/endo/endo/packages/ses/scripts/bundle.js:28:46)
```
where the immediately relevant code in bundle.js is
```js
  const { code: terse } = terser.minify(bundle, {
    mangle: false,
    keep_classnames: true,
  });

  console.log(`Bundle size: ${bundle.length} bytes`);
  console.log(`Minified bundle size: ${terse.length} bytes`);
```

However, this PR itself is clearly completely independent of bundling. I don't see how it could be affected. Any ideas?

gibson042 commented on packages/ses/src/error/note-log-args.js:
```diff
@@ -0,0 +1,216 @@
+    if (cell !== undefined) {
+      cell.value = value;
+    }
+    if (count >= budget) {
+      const condemned = head.prev;
+      spliceOut(condemned); // Drop least recently used
+      map.delete(condemned.key);
+      count -= 1;
+    }
```
This won't work when `budget` is 0, which there are a few ways to fix.
* Require `budget` to be positive.
* Silently promote `budget` to be positive.
* Honor 0 by making `set` noöp.

erights replied:
Good point! I missed that completely.

gibson042 commented on packages/ses/src/error/note-log-args.js:
```diff
@@ -0,0 +1,216 @@
+  const set = (key, value) => {
+    let cell = touchCell(key);
+    if (cell !== undefined) {
+      cell.value = value;
```
This seems to be missing either an early `return` or a following `else`, without which a new cell will be created and added.

erights replied:
Good catch. Fixed.

gibson042 reviewed (commented):
> The unbounded leak that caused us to turn deep stacks off was really the unbounded memory of error-note-graphs.

Huh, really? I would have expected `takeNoteLogArgsArray` to free that... was nothing calling it?

erights commented on packages/ses/src/error/note-log-args.js:
```diff
@@ -0,0 +1,216 @@
+      return undefined;
+    }
+    // Becomes most recently used
+    spliceOut(cell);
+    spliceAfter(head, cell);
```
constant time LRU. Is there a way to do that with an array?

gibson042 replied:
> constant time LRU. Is there a way to do that with an array?

Since the maximum size of the array would be bounded by `budget`, the answer to this question is _technically_ "yes". And it's possible that **`cache = [cache[i], ...cache.slice(0, i), ...cache.slice(i + 1)]`**[^1] or **`cache.splice(0, i + 1, cache[i], ...cache.slice(0, i))`**[^2] is faster than **`spliceOut(cell); spliceAfter(head, cell)`** in absolute terms for the budget we set even though it scales in a worse algorithmic time complexity class, or that it's slower but the difference doesn't matter so we could still go with simpler code.

[^1]: suitably protected against environment manipulation, e.g. `cache = [cache[i], ...arrayIterator(arraySlice(cache, 0, i)), ...arrayIterator(arraySlice(cache, i + 1))]`
[^2]: likewise (e.g., `arraySplice(cache, 0, i + 1, cache[i], ...arrayIterator(arraySlice(cache, 0, i)))`), and also with batching if `budget` can exceed maximum argument count

erights replied:
I agree that is possible. Since it is a fully encapsulated decision, if measurements justify a change, we can always change it later.

Resolving.

erights commented on packages/ses/src/error/note-log-args.js:
```diff
@@ -0,0 +1,216 @@
+  const map = new Map();
+  let count = 0; // count must remain <= budget
+  // As a sigil, `head` uniquely is not in the `map`.
+  const head = makeSelfCell(undefined, undefined);
```
Good point. Fixed.

erights commented:
> > The unbounded leak that caused us to turn deep stacks off was really the unbounded memory of error-note-graphs.
> 
> Huh, really? I would have expected `takeNoteLogArgsArray` to free that... was nothing calling it?

Only for errors that are actually reported. The fatal leak were for errors that could eventually be potentially reported, but were not.

erights commented:
PTAL

gibson042 commented on packages/ses/src/error/note-log-args.js:
```diff
@@ -0,0 +1,219 @@
+  }
+  /** @type {Map<object, DoublyLinkedCell>} */
+  const map = new Map();
+  let count = 0; // count must remain <= budget
```
Nit:
```suggestion
  let size = 0; // size must remain <= budget
```

erights replied:
Done

erights commented:
I also changed the default budget from `10_000` to `1000` because, in the absence of measurement, I'd really like to avoid guessing too high. If too low, it is still better than the status quo, with deep stacks turned off. And we can always increase it later.

We should also consider configuring the budget with an environment option.

erights commented:
Worked around stale tooling. Bug at https://github.com/endojs/endo/issues/1514 . (@kriskowal , thanks!)

erights commented:
Re https://github.com/endojs/endo/pull/1513#issuecomment-1474560497

I reduced the note-args budget again, this time to `100`. At `1000` I ran into one problem:

```js
// Bizarrely, if set to 1000 or 10_000, captp/test/test-gc.js
// test 'test loopback gc' fails ONLY ON WINDOWS.
// TODO: Figure out why, and either special case or fix.
// TODO: Once captp/test/test-gc.js doesn't stop us,
// set this default back to a more reasonable number.
// TODO: make this default configurable by an environment option.
const defaultLogArgsBudget = 100;
```

erights commented:
Given the nature of this PR, before merging, it was checking the effect on performance. The following first crude test is consistent enough to show we're ok, or even a bit better!

https://github.com/endojs/endo/pull/1515 , identical to current master, has the following timings from CI

<img width="1168" alt="image" src="https://user-images.githubusercontent.com/273868/226080693-124bebc0-fade-42f5-9c64-b7d860ca3a30.png">
<img width="1168" alt="image" src="https://user-images.githubusercontent.com/273868/226080750-e35d31f5-5da6-4891-8ad1-7e3bee316238.png">

Compare with the timings of this PR under CI. Bizarrely, this PR is always the same or faster. (ignoring the failed test)

<img width="1165" alt="image" src="https://user-images.githubusercontent.com/273868/226080771-d38dc961-ff69-4114-9eb5-964dbbc089ca.png">
<img width="1170" alt="image" src="https://user-images.githubusercontent.com/273868/226080710-453dba1f-079f-4a38-9045-3b54ce741c8d.png">

erights commented:
At `100` the same test now fails or passes intermittently, which is even worse!

erights commented:
Re https://github.com/endojs/endo/pull/1513#issuecomment-1474626631 , @1516 fixed the immediate problem, the OS-dependent behavior. However, this PR still changes the counts, for reasons we (@michaelfig and I) do not yet precisely understand but believe are fine for now.

@mfig, I changed those lines as we discussed but also added a TODO

```js
  // TODO(mfig,erights): explain why #1513 changed these counts from 3 to 2
  t.is(getFarStats().sendCount.CTP_DROP, 2);
  t.is(getNearStats().recvCount.CTP_DROP, 2);
```
